<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Giuliano Colosimo">
<meta name="dcterms.date" content="2023-11-21">
<meta name="description" content="An example with microsatellite data and R">

<title>Evolvo Ergo Sum - A test of HWE using permutations (to be completed)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../.././figures/avatar.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

<script type="text/javascript">

(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
ga('create', 'UA-24215497', 'auto');

ga('send', {
  hitType: 'pageview',
  'anonymizeIp': true,
});
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta property="og:title" content="Evolvo Ergo Sum - A test of HWE using permutations (to be completed)">
<meta property="og:description" content="An example with microsatellite data and R">
<meta property="og:image" content="https://giulianocolosimo.github.io/EvolvoErgoSum/posts/chisq_permutation/index_files/figure-html/fig-chisq-1.png">
<meta property="og:site-name" content="Evolvo Ergo Sum">
<meta property="og:locale" content="en_US">
<meta property="og:image:height" content="960">
<meta property="og:image:width" content="1344">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Evolvo Ergo Sum</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../aboutme.html" rel="" target="">
 <span class="menu-text">About Me</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-publications" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Publications</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-publications">    
        <li>
    <a class="dropdown-item" href="../../peerreviewed.html" rel="" target="">
 <span class="dropdown-text">Peer-reviewed</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../nonpeerreviewed.html" rel="" target="">
 <span class="dropdown-text">Non Peer-reviewed</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-projects" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Projects</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-projects">    
        <li>
    <a class="dropdown-item" href="../../taco.html" rel="" target="">
 <span class="dropdown-text">T. A. Co.</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html" rel="" target="">
 <span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../photos.html" rel="" target="">
 <span class="menu-text">Photos</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../contacts.html" rel="" target="">
 <span class="menu-text">Contacts</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">A test of HWE using permutations (to be completed)</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
      <a href="../../index.html" class="sidebar-logo-link">
      <img src="../.././figures/avatar.jpg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      <div class="sidebar-tools-main tools-wide">
    <a href="https://www.linkedin.com/in/giulianocolosimo/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-linkedin"></i></a>
    <a href="https://github.com/GiulianoColosimo" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
    <a href="mailto:giuliano.colosimo@uniroma2.it" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-envelope"></i></a>
    <a href="https://drive.proton.me/urls/XTZG4M52RR#mpcdheaxnXE8" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-file-pdf"></i></a>
</div>
      </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#data-wrangling" id="toc-data-wrangling" class="nav-link" data-scroll-target="#data-wrangling">Data wrangling</a></li>
  <li><a href="#next-step-coming-soon" id="toc-next-step-coming-soon" class="nav-link" data-scroll-target="#next-step-coming-soon">Next step coming soon…</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A test of HWE using permutations (to be completed)</h1>
  <div class="quarto-categories">
    <div class="quarto-category">R</div>
    <div class="quarto-category">Chi square</div>
    <div class="quarto-category">Permutation</div>
    <div class="quarto-category">Statistics</div>
  </div>
  </div>

<div>
  <div class="description">
    An example with microsatellite data and R
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Giuliano Colosimo </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 21, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>Working with microsatellite genotypes data for many years I always wanted to fully understand how to build a test to verify whether the population(s) under analysis is(are) in Hardy-Weinberg (HW) equilibrium <span class="citation" data-cites="Hardy1908">(<a href="#ref-Hardy1908" role="doc-biblioref">Hardy, 1908</a>)</span> (or, to be more precise, if the population’s genotypes reflects HW proportions). Checking for HW proportions is generally the first step in any kind of population genetic analysis <span class="citation" data-cites="Waples2019">(<a href="#ref-Waples2019" role="doc-biblioref">Waples <em>et al.</em>, 2018</a>)</span>.</p>
<p>One of the most basic approach to perform such a test is to use the <span class="math inline">\(\chi^{2}\)</span> method <span class="citation" data-cites="Allendorf2022">(<a href="#ref-Allendorf2022" role="doc-biblioref">Allendorf <em>et al.</em>, 2022</a>)</span>. This approach is particularly useful when the number of loci and alleles is too big to perform a Fisher’s Exact Test <span class="citation" data-cites="Fisher1935">(<a href="#ref-Fisher1935" role="doc-biblioref">Fisher, 1935</a>)</span>. The straight up formula of the <span class="math inline">\(\chi^{2}\)</span> test is relatively simple: <span class="math inline">\(\chi^{2} = \sum\frac{(OBSERVED-EXPECTED)^2}{EXPECTED}\)</span>, where OBSERVED and EXPECTED refer to the number of observed and expected genotypes for each locus and allele. The <span class="math inline">\(\chi^{2}\)</span> value obtained is then compared to a set of values obtained under the assumption of no violation of the <span class="math inline">\(H_0\)</span>.</p>
<div class="cell">
<details>
<summary>Show the R code used to produce the graph</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>chiTestVal <span class="ot">&lt;-</span> </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="fl">0.001</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">pchisq</span>(chiTestVal, <span class="at">df =</span> <span class="dv">1</span>), chiTestVal, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="fu">expression</span>(chi<span class="sc">^</span><span class="dv">2</span><span class="sc">*</span><span class="st">"distribution"</span>),</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Test Statistic Value"</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"p-value"</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">pchisq</span>(chiTestVal, <span class="at">df =</span> <span class="dv">2</span>), chiTestVal, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">pchisq</span>(chiTestVal, <span class="at">df =</span> <span class="dv">3</span>), chiTestVal, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"green"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">pchisq</span>(chiTestVal, <span class="at">df =</span> <span class="dv">4</span>), chiTestVal, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"brown"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">pchisq</span>(chiTestVal, <span class="at">df =</span> <span class="dv">5</span>), chiTestVal, <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"deeppink"</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="fl">0.05</span>, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, </span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"DoF = 1"</span>, <span class="st">"DoF = 2"</span>, <span class="st">"DoF = 3"</span>, <span class="st">"DoF = 4"</span>, <span class="st">"DoF = 5"</span>), </span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"brown"</span>, <span class="st">"deeppink"</span>), <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-chisq" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="index_files/figure-html/fig-chisq-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;1: <span class="math inline">\(\chi^{2}\)</span> distribution graph showing the test statistic on the y axis and its associated p-value on the x axis. The test statistic is calculated with different Degrees of Freedom (DoF). The red dashed line marks the widely used threshold of p-value = 0.05 under which we can reject <span class="math inline">\(H_0\)</span>. As DoF increases, the test statistic value needed to reject <span class="math inline">\(H_0\)</span> gets higher.</figcaption>
</figure>
</div>
</div>
</div>
<p>While the <span class="math inline">\(\chi^{2}\)</span> test is applied in many different situations, there are some cases, for example when there are many genotypes with low expected numbers due the presence of many different alleles, where a different approach may serve us better in testing <span class="math inline">\(H_0\)</span>: permutation tests <span class="citation" data-cites="Allendorf2022">(<a href="#ref-Allendorf2022" role="doc-biblioref">Allendorf <em>et al.</em>, 2022</a>)</span>.</p>
<p>The overall idea behind this approach is to recreate X number of randomized datasets starting from the observed data. For each generated dataset a <span class="math inline">\(\chi^{2}\)</span> value is then calculated and compared with the value obtained from the original dataset. The prorportion of randomized dataset that produce a <span class="math inline">\(\chi^{2}\)</span> value as large as or lager than the observed data give us an unbiased estimate of the probability of rejecting <span class="math inline">\(H_0\)</span> <span class="citation" data-cites="Allendorf2022">(<a href="#ref-Allendorf2022" role="doc-biblioref">Allendorf <em>et al.</em>, 2022</a>)</span>.</p>
<p>In this post I will show how I build a permutation test procedure to evaluate the probability if rejecting <span class="math inline">\(H_0\)</span> when working with microsatellite genotypes. I will be using a microsatellite dataset available from Dryad (click <a href="https://drive.proton.me/urls/W9206FWP1C#LlZJdXqsWypn">here</a> to download). The dataset is borrowed from <span class="citation" data-cites="Chan2020">Chan <em>et al.</em> (<a href="#ref-Chan2020" role="doc-biblioref">2020</a>)</span> and consists of 27 microsatellites typed in 237 individuals. The original dataset is available in <strong><font color="green"> genpop </font></strong> format. This is a widely used format in R applications. The procedure illustrated here works directly with matrices. Therefore in this first part of the post I will cover the step-by-step procedure to transition from <strong><font color="green"> genpop </font></strong> to <strong><font color="green"> matrix </font></strong>.</p>
</section>
<section id="data-wrangling" class="level1">
<h1>Data wrangling</h1>
<p>First thing to do is to load the data file using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet)                                    <span class="co"># package to load</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sceloporus_data <span class="ot">&lt;-</span> <span class="fu">read.genepop</span>(<span class="st">"path/to/file.gen"</span>,  <span class="co"># specify the path to the genepop file</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">ncode =</span> <span class="dv">3</span>,           <span class="co"># how many digits are used for alleles</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">quiet =</span> T)           <span class="co"># print loading progress on screen</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <em>read.genepop()</em> function from <strong>adegenet</strong> <span class="citation" data-cites="Jombart2008a">(<a href="#ref-Jombart2008a" role="doc-biblioref">Jombart, 2008</a>)</span> automatically converts the data in to a <strong><font color="green"> genind </font></strong> object, the data format used by <strong>adegenet</strong>. We can take a look at what it is like just by typing the name of data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>sceloporus_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/// GENIND OBJECT /////////

 // 237 individuals; 27 loci; 442 alleles; size: 499.2 Kb

 // Basic content
   @tab:  237 x 442 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 3-35)
   @loc.fac: locus factor for the 442 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: read.genepop(file = "~/Documents/websites/EvolvoErgoSum/data/doi_10.5061_dryad.sxksn0316__v2/microsatellite_genotypes.gen", 
    ncode = 3, quiet = T)

 // Optional content
   @pop: population of each individual (group size range: 6-76)</code></pre>
</div>
</div>
<p>This data format is quite useful as it gives a first glance of what is that we are dealing with. Unfortunately this is not the format needed to perform this custom test so we first need to transform the data a bit. In fact this permutation procedure takes as input file a matrix of genotypes. Rather than giving you an example right now I think is best if we proceed and transform the dataset so that you can see how the input file should look like. The first step is to transform the data object in a data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sceloporus_data_df <span class="ot">&lt;-</span> <span class="fu">genind2df</span>(sceloporus_data)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>sceloporus_data_df[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              pop sa5220 sa5229 sa6002 sa6012  sar80  sar84 sarms0001 sarms0344
DJL286  TCWC93505 109112 233233 121129 146146 362398 117123    207207    251251
DJL297  TCWC93505 109109 201225 121129 134140 350358 117117    199199    242242
DL924   TCWC93505 109112 185185 121129 146152   &lt;NA&gt; 117117    207207    242245
DL925   TCWC93505 109112 221241 121121 142144 374382 117117    207199    251251
ESP9057 TCWC93505 109109 225221 121121 136136 378378 117117    207207    242242
ESP9070 TCWC93505 109112 205229 121121 140144 374374 117117    207199    251242
ESP9479 TCWC93505 109109 229237 121121 146144 350354 117135    207207    251251
ESP9480 TCWC93505 109109 201229 121121 134128 354353 117135    207207    251242
ESP9485 TCWC93505 109112 233225 121121 128138 350353 117135    207207    251251
ESP9486 TCWC93505 109112 225245 129117 140140 350382 117131    207207    242242</code></pre>
</div>
</div>
<p>By inspecting the first 10 individuals and 9 loci we notice that the function has nicely organized our individuals in rows, kept the population information in the first column, and used the other columns for the genotypes. This format is very close to what we need, but not quite yet. First, we change the coding scheme of missing genotypes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sceloporus_data_df[<span class="fu">apply</span>(sceloporus_data_df, <span class="dv">2</span>, is.na)] <span class="ot">&lt;-</span> <span class="st">"000000"</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sceloporus_data_df[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              pop sa5220 sa5229 sa6002 sa6012  sar80  sar84 sarms0001 sarms0344
DJL286  TCWC93505 109112 233233 121129 146146 362398 117123    207207    251251
DJL297  TCWC93505 109109 201225 121129 134140 350358 117117    199199    242242
DL924   TCWC93505 109112 185185 121129 146152 000000 117117    207207    242245
DL925   TCWC93505 109112 221241 121121 142144 374382 117117    207199    251251
ESP9057 TCWC93505 109109 225221 121121 136136 378378 117117    207207    242242
ESP9070 TCWC93505 109112 205229 121121 140144 374374 117117    207199    251242
ESP9479 TCWC93505 109109 229237 121121 146144 350354 117135    207207    251251
ESP9480 TCWC93505 109109 201229 121121 134128 354353 117135    207207    251242
ESP9485 TCWC93505 109112 233225 121121 128138 350353 117135    207207    251251
ESP9486 TCWC93505 109112 225245 129117 140140 350382 117131    207207    242242</code></pre>
</div>
</div>
<p>Just a few more lines of code to transform everything from a data frame to a matrix and retain all the information that may be useful down the road.</p>
<div class="cell">
<details>
<summary>Show the R code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sceloporus_data_mt <span class="ot">&lt;-</span> <span class="fu">apply</span>(sceloporus_data_df[,<span class="sc">-</span><span class="dv">1</span>], <span class="dv">2</span>, <span class="cf">function</span>(x){</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(<span class="fu">substr</span>(x, <span class="dv">1</span>, <span class="dv">3</span>), <span class="fu">substr</span>(x, <span class="dv">4</span>, <span class="dv">6</span>), <span class="at">sep =</span> <span class="st">"/"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>sceloporus_data_mt <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.character</span>(sceloporus_data_df[,<span class="dv">1</span>]), sceloporus_data_mt)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sceloporus_data_mt) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sceloporus_data_df)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>sceloporus_data_mt[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                    sa5220    sa5229    sa6002    sa6012    sar80     sar84    
DJL286  "TCWC93505" "109/112" "233/233" "121/129" "146/146" "362/398" "117/123"
DJL297  "TCWC93505" "109/109" "201/225" "121/129" "134/140" "350/358" "117/117"
DL924   "TCWC93505" "109/112" "185/185" "121/129" "146/152" "000/000" "117/117"
DL925   "TCWC93505" "109/112" "221/241" "121/121" "142/144" "374/382" "117/117"
ESP9057 "TCWC93505" "109/109" "225/221" "121/121" "136/136" "378/378" "117/117"
ESP9070 "TCWC93505" "109/112" "205/229" "121/121" "140/144" "374/374" "117/117"
ESP9479 "TCWC93505" "109/109" "229/237" "121/121" "146/144" "350/354" "117/135"
ESP9480 "TCWC93505" "109/109" "201/229" "121/121" "134/128" "354/353" "117/135"
ESP9485 "TCWC93505" "109/112" "233/225" "121/121" "128/138" "350/353" "117/135"
ESP9486 "TCWC93505" "109/112" "225/245" "129/117" "140/140" "350/382" "117/131"</code></pre>
</div>
</div>
</section>
<section id="next-step-coming-soon" class="level1">
<h1>Next step coming soon…</h1>
</section>
<section id="references" class="level1">




</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Allendorf2022" class="csl-entry" role="listitem">
Allendorf, F.W., Funk, W.C., Aitken, S.N., Byrne, M., Luikart, G. and Antunes, A. 2022. <em><a href="https://doi.org/10.1093/oso/9780198856566.001.0001"><span class="nocase">Conservation and the Genomics of Populations</span></a></em>. Oxford University Press.
</div>
<div id="ref-Chan2020" class="csl-entry" role="listitem">
Chan, L.M., Painter, C.W., Hill, M.T., Hibbitts, T.J., Leavitt, D.J., Ryberg, W.A., <em>et al.</em> 2020. <a href="https://doi.org/10.1371/journal.pone.0238194">Phylogeographic structure of the dunes sagebrush lizard, an endemic habitat specialist</a>. <em><span>PLOS</span> <span>ONE</span></em>, <strong>15</strong>: e0238194. Public Library of Science (<span>PLoS</span>).
</div>
<div id="ref-Fisher1935" class="csl-entry" role="listitem">
Fisher, R.A. 1935. <a href="https://doi.org/10.1111/j.2397-2335.1935.tb04208.x">The logic of inductive inference</a>. <em>Journal of the Royal Statistical Society</em>, <strong>98</strong>: 39–54.
</div>
<div id="ref-Hardy1908" class="csl-entry" role="listitem">
Hardy, G.H. 1908. <a href="https://doi.org/10.1126/science.28.706.49">Mendelian proportions in a mixed population</a>. <em>Science</em>, <strong>28</strong>: 49–50.
</div>
<div id="ref-Jombart2008a" class="csl-entry" role="listitem">
Jombart, T. 2008. <a href="https://doi.org/10.1093/bioinformatics/btn129"><span class="nocase">adegenet: a R package for the multivariate analysis of genetic markers.</span></a> <em>Bioinformatics</em>, <strong>24</strong>: 1403–1405.
</div>
<div id="ref-Waples2019" class="csl-entry" role="listitem">
Waples, R.S., Rus Hoelzel, A., Gaggiotti, O., Tiedemann, R., Palsbøll, P.J., Cipriano, F., <em>et al.</em> 2018. Guidelines for genetic data analysis. <em>Journal of Cetacean Research and Management</em>, <strong>18</strong>: 33–80.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://utteranc.es/client.js" repo="GiulianoColosimo/EvolvoErgoSum" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">© Copyright 2023 Giuliano Colosimo</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This page is built with ❤️ and <a href="https://quarto.org/">Quarto</a>. Inspiration for this site comes from <a href="https://www.mm218.dev">here</a></div>
  </div>
</footer>



</body></html>